File: simulator.js
// Simulator class for running AI-only game matches

class SimulatorGame extends Game {
    constructor(aiPlayers) {
        super();
        this.aiPlayers = aiPlayers;
        this.players = aiPlayers.map(ai => ai.player);
        this.gameStarted = true;
        this.turn = 1;
        this.phase = 1;
        this.silent = true; // Don't log to console during simulation
    }

    // Override log to be silent during simulation
    log(message, type = 'info') {
        // Silent during simulation
    }

    // Override render to avoid DOM updates during simulation
    render() {
        // No rendering during simulation
    }

    playTurn() {
        // Phase 1: Production and Expenses
        // Get actions from all AI players
        for (const aiPlayer of this.aiPlayers) {
            if (!aiPlayer.player.bankrupt) {
                const actions = aiPlayer.getActions(this);
                aiPlayer.executeActions(this, actions);
            }
        }

        // Pay costs for all players
        for (const player of this.players) {
            if (player.bankrupt) continue;

            const costs = player.calculateCosts();
            if (!player.removeTokens(costs)) {
                player.bankrupt = true;
            } else {
                player.calculateProduction();
            }
        }

        // Check game over after phase 1
        if (this.checkGameOverSilent()) {
            return true;
        }

        // Phase 2: Market
        this.processMarketSilent();

        // Phase 3: Taxes
        this.processTaxesSilent();

        // Check game over after phase 3
        if (this.checkGameOverSilent()) {
            return true;
        }

        // Phase 4: Events
        this.processEventSilent();

        // Reset temporary effects for next turn
        for (const player of this.players) {
            player.resetTempEffects();
        }
        this.marketPenalty = 1;

        this.turn++;

        // Max 50 turns
        if (this.turn > 50) {
            return true;
        }

        return false;
    }

    processMarketSilent() {
        // Calculate market size
        const dice1 = Math.floor(Math.random() * 6) + 1;
        const dice2 = Math.floor(Math.random() * 6) + 1;
        const diceTotal = dice1 + dice2;

        let coefficientSum = 0;
        for (const player of this.players) {
            if (!player.bankrupt) {
                const level = LEVELS[player.getLevel()];
                coefficientSum += level.coefficient;
            }
        }

        const activePlayers = this.players.filter(p => !p.bankrupt).length;
        this.marketSize = Math.floor((diceTotal * activePlayers * 2 + coefficientSum * 2) * this.marketPenalty);

        // Calculate total reputation
        let totalReputation = 0;
        for (const player of this.players) {
            if (!player.bankrupt) {
                totalReputation += player.getTotalReputation();
            }
        }

        if (totalReputation === 0) {
            return;
        }

        // Distribute market shares
        for (const player of this.players) {
            if (player.bankrupt) continue;

            const marketShare = (player.getTotalReputation() / totalReputation);
            const potentialSales = Math.floor(this.marketSize * marketShare);
            player.sales = Math.min(potentialSales, player.production);
            player.revenue = player.sales;

            player.addTokens(player.revenue);
        }
    }

    processTaxesSilent() {
        for (const player of this.players) {
            if (player.bankrupt) continue;

            // Calculate taxes
            const revenueTax = Math.floor(player.revenue * 0.2);
            const employeeTax = player.employees * 1;
            const machineTax = player.machines * 2;
            const level = LEVELS[player.getLevel()];
            const structuralTax = level.taxLevel;

            const totalTax = revenueTax + employeeTax + machineTax + structuralTax;

            if (!player.removeTokens(totalTax)) {
                player.bankrupt = true;
            }
        }
    }

    processEventSilent() {
        const activePlayers = this.players.filter(p => !p.bankrupt);

        // Apply 1 event per player
        for (const player of activePlayers) {
            const random = Math.random();
            let eventType, eventList;

            if (random < 0.4) {
                eventType = 'BONUS';
                eventList = EVENTS.BONUS;
            } else if (random < 0.8) {
                eventType = 'MALUS';
                eventList = EVENTS.MALUS;
            } else {
                eventType = 'CHOICE';
                eventList = EVENTS.CHOICE;
            }

            const event = eventList[Math.floor(Math.random() * eventList.length)];

            // Apply event to player or game
            if (event.effect === 'malus' && event.name === 'Crise') {
                event.apply(this);
            } else {
                event.apply(player);
            }
        }
    }

    checkGameOverSilent() {
        // Check for 100 gold victory
        for (const player of this.players) {
            if (player.tokens.gold >= 100) {
                return true;
            }
        }

        // Check for last survivor
        const activePlayers = this.players.filter(p => !p.bankrupt);
        if (activePlayers.length <= 1) {
            return true;
        }

        return false;
    }
}

class MatchData {
    constructor(matchNumber) {
        this.matchNumber = matchNumber;
        this.data = [];
    }

    recordTurn(game) {
        for (const player of game.players) {
            this.data.push({
                match: this.matchNumber,
                turn: game.turn,
                player: player.name,
                jetons: player.getTotalTokens(),
                jetons_bronze: player.tokens.bronze,
                jetons_silver: player.tokens.silver,
                jetons_gold: player.tokens.gold,
                reputation: player.reputation,
                employees: player.employees,
                machines: player.machines,
                stock: player.production,
                sales: player.sales,
                revenue: player.revenue,
                bankrupt: player.bankrupt ? 1 : 0,
                level: player.getLevel()
            });
        }
    }

    toCSVRows() {
        return this.data;
    }
}

class Simulator {
    constructor() {
        this.results = [];
    }

    async runSimulation(numMatches, onProgress = null) {
        this.results = [];

        for (let i = 0; i < numMatches; i++) {
            const matchData = this.runMatch(i + 1);
            this.results.push(...matchData.toCSVRows());

            // Call progress callback
            if (onProgress) {
                onProgress(i + 1, numMatches);
            }

            // Yield to browser every 10 matches to avoid blocking
            if (i % 10 === 9) {
                await new Promise(resolve => setTimeout(resolve, 0));
            }
        }

        return this.results;
    }

    runMatch(matchNumber) {
        // Create 4 AI players with different strategies (newly instantiated each match for variability)
        const strategies = [
            new BalancedStrategy(),
            new AggressiveStrategy(),
            new ConservativeStrategy(),
            new ReputationFocusedStrategy()
        ];

        // Shuffle strategies to randomize player assignments each match
        this.shuffleArray(strategies);

        const aiPlayers = strategies.map((strategy, index) => {
            const player = new Player(index, strategy.name);
            return new AIPlayer(player, strategy);
        });

        const game = new SimulatorGame(aiPlayers);
        const matchData = new MatchData(matchNumber);

        // Record initial state
        matchData.recordTurn(game);

        // Play up to 50 turns
        let gameOver = false;
        while (!gameOver && game.turn <= 50) {
            gameOver = game.playTurn();
            matchData.recordTurn(game);
        }

        return matchData;
    }

    // Fisher-Yates shuffle algorithm
    shuffleArray(array) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
        return array;
    }

    toCSV(results) {
        const headers = [
            'match', 'turn', 'player', 'jetons', 'jetons_bronze', 'jetons_silver',
            'jetons_gold', 'reputation', 'employees', 'machines', 'stock',
            'sales', 'revenue', 'bankrupt', 'level'
        ];

        const rows = [headers.join(',')];

        for (const row of results) {
            const values = [
                row.match,
                row.turn,
                row.player,
                row.jetons,
                row.jetons_bronze,
                row.jetons_silver,
                row.jetons_gold,
                row.reputation,
                row.employees,
                row.machines,
                row.stock,
                row.sales,
                row.revenue,
                row.bankrupt,
                row.level
            ];
            rows.push(values.join(','));
        }

        return rows.join('\n');
    }

    downloadCSV(results, filename) {
        const csvContent = this.toCSV(results);
        const blob = new Blob([csvContent], { type: 'text/csv' });
        const url = window.URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = filename;
        document.body.appendChild(a);
        a.click();
        document.body.removeChild(a);
        window.URL.revokeObjectURL(url);
    }
}

File: backend/simulator.py
"""
Match simulation engine with CSV export
"""
import csv
from io import StringIO
from typing import List, Dict, Any
from datetime import datetime
from .game_logic import Game, Player
from .ai_player import create_ai_players, AIPlayer


class MatchData:
    """Stores data for a single match"""

    def __init__(self, match_number: int):
        self.match_number = match_number
        self.turns_data: List[Dict[str, Any]] = []

    def add_turn(self, turn: int, players: List[Player]):
        """Record player states for a turn"""
        for player in players:
            self.turns_data.append({
                'match': self.match_number,
                'turn': turn,
                'player': player.name,
                'jetons': player.tokens.total(),
                'jetons_bronze': player.tokens.bronze,
                'jetons_silver': player.tokens.silver,
                'jetons_gold': player.tokens.gold,
                'reputation': player.reputation,
                'employees': player.employees,
                'machines': player.machines,
                'stock': player.production,  # Production is the stock produced
                'sales': player.sales,
                'revenue': player.revenue,
                'bankrupt': player.bankrupt,
                'level': player.get_level().value
            })


class SimulationResult:
    """Contains results of multiple matches"""

    def __init__(self):
        self.matches: List[MatchData] = []
        self.start_time: str = ""
        self.end_time: str = ""
        self.total_matches: int = 0

    def add_match(self, match: MatchData):
        """Add a match to results"""
        self.matches.append(match)

    def to_csv(self) -> str:
        """Convert simulation results to CSV format"""
        output = StringIO()
        fieldnames = [
            'match',
            'turn',
            'player',
            'jetons',
            'jetons_bronze',
            'jetons_silver',
            'jetons_gold',
            'reputation',
            'employees',
            'machines',
            'stock',
            'sales',
            'revenue',
            'bankrupt',
            'level'
        ]

        writer = csv.DictWriter(output, fieldnames=fieldnames)
        writer.writeheader()

        for match in self.matches:
            for turn_data in match.turns_data:
                writer.writerow(turn_data)

        return output.getvalue()

    def get_summary(self) -> Dict[str, Any]:
        """Get summary statistics"""
        if not self.matches:
            return {}

        # Count wins per strategy
        winners = {}
        total_turns = []

        for match in self.matches:
            # Find winner (last turn, non-bankrupt player with most tokens)
            if match.turns_data:
                last_turn = max(t['turn'] for t in match.turns_data)
                last_turn_data = [t for t in match.turns_data if t['turn'] == last_turn]

                # Find winner
                active_players = [t for t in last_turn_data if not t['bankrupt']]
                if active_players:
                    winner = max(active_players, key=lambda x: x['jetons'])
                    winners[winner['player']] = winners.get(winner['player'], 0) + 1

                total_turns.append(last_turn)

        return {
            'total_matches': len(self.matches),
            'average_turns': sum(total_turns) / len(total_turns) if total_turns else 0,
            'wins_by_player': winners,
            'start_time': self.start_time,
            'end_time': self.end_time
        }


class Simulator:
    """Runs multiple matches with AI players"""

    def __init__(self, max_turns: int = 50):
        self.max_turns = max_turns

    def run_match(self, match_number: int) -> MatchData:
        """Run a single match"""
        # Create game with AI players
        game = Game(num_players=4)
        ai_players = create_ai_players(num_players=4)

        # Replace game players with AI players' player objects
        game.players = [ai.player for ai in ai_players]

        # Create match data recorder
        match_data = MatchData(match_number)

        # Record initial state (turn 0)
        match_data.add_turn(0, game.players)

        # Play turns
        for turn in range(1, self.max_turns + 1):
            # Get actions from each AI
            all_actions = {}
            for ai in ai_players:
                if not ai.player.bankrupt:
                    actions = ai.get_actions(game)
                    all_actions[ai.player.name] = actions

            # Play the turn
            game_continues = game.play_turn(all_actions)

            # Record turn data
            match_data.add_turn(turn, game.players)

            # Check if game is over
            if not game_continues:
                break

        return match_data

    def run_simulation(self, num_matches: int, progress_callback=None) -> SimulationResult:
        """
        Run multiple matches
        progress_callback: Optional function called with (current, total) after each match
        """
        result = SimulationResult()
        result.total_matches = num_matches
        result.start_time = datetime.now().isoformat()

        for i in range(num_matches):
            match_data = self.run_match(i + 1)
            result.add_match(match_data)

            if progress_callback:
                progress_callback(i + 1, num_matches)

        result.end_time = datetime.now().isoformat()
        return result


def simulate_matches(num_matches: int) -> str:
    """
    Convenience function to simulate matches and return CSV
    """
    simulator = Simulator(max_turns=50)
    result = simulator.run_simulation(num_matches)
    return result.to_csv()
